src/libcore/va_list.rs

```diff
@@ -0,0 +1,97 @@
+// Copyright 2018 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+#![unstable(feature = "c_variadic",
+            reason = "the `c_variadic` feature has not been properly tested on \
+                      all supported platforms",
+            issue = "27745")]
+
+//! Implementation of a `va_list`.
+
+#[cfg(target_arch = "aarch64")]
+#[repr(C)]
+#[derive(Debug)]
+/// AArch64 ABI implementation of a `va_list`. See the
+/// [Aarch64 Procedure Call Standard] for more details.
+///
+/// [AArch64 Procedure Call Standard]:
+/// http://infocenter.arm.com/help/topic/com.arm.doc.ihi0055b/IHI0055B_aapcs64.pdf
+pub struct VaListTag {
+    stack: *mut (),
+    gr_top: *mut (),
+    vr_top: *mut (),
+    gr_offs: i32,
+    vr_offs: i32,
+}
+
+#[cfg(any(target_arch = "arm", target_arch = "mips", target_arch = "powerpc64",
+          target_arch = "x86", windows))]
+/// Basic byte pointer variant of a `va_list`.
+pub type VaListTag = *const i8;
+
+#[cfg(any(target_arch = "mips"))]
+/// Basic void pointer variant of a `va_list`.
+pub type VaListTag = *const ();
+
+#[cfg(target_arch = "powerpc")]
+#[repr(C)]
+#[derive(Debug)]
+/// PowerPC ABI implementation of a `va_list`.
+pub struct VaListTag {
+    gpr: u8,
+    fpr: u8,
+    reserved: u16,
+    overflow_arg_area: *mut (),
+    reg_save_area: *mut (),
+}
+
+#[cfg(all(target_arch = "x86_64", not(windows)))]
+#[repr(C)]
+#[derive(Debug)]
+/// x86_64 ABI implementation of a `va_list`.
+pub struct VaListTag {
+    gp_offset: i32,
+    fp_offset: i32,
+    overflow_arg_area: *mut (),
+    reg_save_area: *mut (),
+}
+
+/// A wrapper for a `va_list`
+#[cfg(not(stage0))]
+#[lang = "va_list"]
+#[unstable(feature = "c_variadic",
+           reason = "the `c_variadic` feature has not been properly tested on \
+                     all supported platforms",
+           issue = "27745")]
+#[derive(Debug)]
+pub struct VaList<'a>(&'a mut VaListTag);
```

Comment:
Wait, if it's always a mutable reference to some `VaListTag`, why would we need to know the shape of `VaListTag`? Don't the LLVM intrinsics already know how to do everything here?

Reply:
> why would we need to know the shape of VaListTag?

In the LLVM Lang Ref

> The LLVM assembly language reference manual does not define what this type is...
[1]

After looking through the llvm code it seems that the compiler must define this type (clang does this too). I could be mistaken though.

This is also an improvement on my previous revisions, as we now get better debug info (field names etc).

The reason why I implemented this as `VaList(&mut VaListTag)` instead of essentially just the `VaListTag` is due to the following.

clang emits something like the following for functions that take a `va_list`.

```
; Given uint64_t va_list_test(uint64_t, va_list) clang emits
define u64 @va_list_test(u64, %struct.__va_list_tag*) ...
```

Rather than making the interface for `c_variadic` functions in rust like the following

```rust
pub unsafe extern "C" fn some_fn(something: Type, ap: &mut VaList) -> OtherType
```

I wanted the interface to be

```rust
pub unsafe extern "C" fn some_fn(something: Type, mut ap: VaList) -> OtherType
```

Which is why I made the `VaList` type always contain a single `&mut VaListTag` member.

[1]: https://llvm.org/docs/LangRef.html#variable-argument-handling-intrinsics

Reply:
You'll have to add `#[repr(transparent)]` to make sure that the newtype of a pointer gets the same call ABI as the pointer itself would in C.

Reply:
Good catch